File: predictor_api/predictor_api/service/group_service.py
from typing import Any
from uuid import UUID

from fastapi import HTTPException

from db_handler.db_handler.model.column import Column
from db_handler.db_handler.model.order_by import OrderBy
from db_handler.db_handler.model.query_condition import QueryCondition
from db_handler.db_handler.model.query_condition_group import (
    QueryConditionGroup
)
from db_handler.db_handler.model.query_request import QueryRequest
from db_handler.db_handler.model.query_response import QueryResponse
from db_handler.db_handler.model.table import Table
from db_handler.db_handler.model.table_join import TableJoin
from db_handler.db_handler.model.type.condition_operator import (
    ConditionOperator
)
from db_handler.db_handler.model.type.sql_operator import SqlOperator
from db_handler.db_handler.model.type.table_join_type import TableJoinType
from db_handler.db_handler.model.update_request import UpdateRequest
from db_handler.db_handler.service.database_query_service import (
    DatabaseQueryService
)
from db_handler.db_handler.util.store_constants import StoreConstants
from predictor_api.predictor_api.model.group import Group
from predictor_api.predictor_api.model.group_update import GroupUpdate
from predictor_api.predictor_api.model.league_template import LeagueTemplate
from predictor_api.predictor_api.model.team import Team
from predictor_api.predictor_api.model.tournament import Tournament
from predictor_api.predictor_api.model.tournament_template import (
    TournamentTemplate
)
from predictor_api.predictor_api.service.tournament_service import (
    TournamentService
)
from predictor_api.predictor_api.util.predictor_constants import (
    PredictorConstants
)


class GroupService:
    """
    Service for performing group-related actions.

    Attributes:
        __query_service (DatabaseQueryService): The database query service.
        __tournament_service (TournamentService): The tournament service.
    """

    def __init__(
            self,
            database_query_service: DatabaseQueryService,
            tournament_service: TournamentService) -> None:
        """
        Initialise the GroupService.

        Args:
            database_query_service (DatabaseQueryService): The database query
                service.
            tournament_service (TournamentService): The tournament service.
        """
        self.__query_service = database_query_service
        self.__tournament_service = tournament_service

    def get_groups(self, tournament_id: UUID) -> list[Group]:
        """
        Retrieve stored groups.

        Args:
            tournament_id (UUID): The id of the tournament whose groups are
                to be retrieved.

        Returns:
            list[Group]: The stored groups.
        """
        self.__tournament_service.get_tournament_by_id(tournament_id)

        if not self.__tournament_has_group_stage(tournament_id):
            raise HTTPException(
                status_code=404,
                detail="The tournament with the supplied id does not have a "
                       "group stage."
            )

        return self.__retrieve_and_build_groups(tournament_id)

    def update_groups(
            self,
            tournament_id: UUID,
            groups: list[GroupUpdate]) -> list[Group]:
        """
        Update existing groups.

        Args:
            tournament_id (UUID): The id of the tournament whose groups are
                to be updated.
            groups (list[GroupUpdate]): The groups to update.

        Returns:
            list[Group]: The newly updated groups.
        """
        self.__tournament_service.get_tournament_by_id(tournament_id)

        if not self.__tournament_has_group_stage(tournament_id):
            raise HTTPException(
                status_code=404,
                detail="The tournament with the supplied id does not have a "
                       "group stage."
            )

        self.__query_service.update_records(
            UpdateRequest(
                operation=SqlOperator.UPDATE,
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    Group.get_target_table(tournament_id)
                ),
                records=list(
                    map(
                        lambda group:
                        group.model_dump(exclude_none=True),
                        groups
                    )
                )
            )
        )

        return self.__retrieve_and_build_groups(
            tournament_id,
            QueryConditionGroup.of(
                QueryCondition(
                    column=Column.of("group", StoreConstants.ID),
                    operator=ConditionOperator.IN,
                    value=list(
                        map(
                            lambda group:
                            group.id,
                            groups
                        )
                    )
                )
            )
        )

    def get_group_by_id(self, tournament_id: UUID, group_id: UUID) -> Group:
        """
        Retrieve a single stored group by its id.

        Args:
            tournament_id (UUID): The id of the tournament the group belongs
                to.
            group_id (UUID): The id of the group to retrieve.

        Returns:
            Group: The retrieved group.
        """
        self.__tournament_service.get_tournament_by_id(tournament_id)

        if not self.__tournament_has_group_stage(tournament_id):
            raise HTTPException(
                status_code=404,
                detail="The tournament with the supplied id does not have a "
                       "group stage."
            )

        group_response: QueryResponse = self.__query_service.retrieve_records(
            QueryRequest(
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    Group.get_target_table(tournament_id)
                ),
                conditionGroup=QueryConditionGroup.of(
                    QueryCondition.of(
                        Column.of(StoreConstants.ID),
                        group_id
                    )
                )
            )
        )

        if len(group_response.records) == 0:
            raise HTTPException(
                status_code=404,
                detail="No groups found with a matching id."
            )

        return self.__retrieve_and_build_groups(
            tournament_id,
            QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of("group", StoreConstants.ID),
                    group_id
                )
            )
        )[0]

    def add_teams_to_group(
            self,
            tournament_id: UUID,
            group_id: UUID,
            team_ids: list[UUID]) -> Group:
        """
        Add teams to a group.

        Args:
             tournament_id (UUID): The id of the tournament the group belongs
                to.
             group_id (UUID): The id of the group to add teams to.
             team_ids (list[UUID]): The ids of the teams to add to the group.

        Returns:
            Group: The updated group.
        """
        self.__tournament_service.get_tournament_by_id(tournament_id)

        if not self.__tournament_has_group_stage(tournament_id):
            raise HTTPException(
                status_code=404,
                detail="The tournament with the supplied id does not have a "
                       "group stage."
            )

        group_response: QueryResponse = self.__query_service.retrieve_records(
            QueryRequest(
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    Group.get_target_table(tournament_id)
                ),
                conditionGroup=QueryConditionGroup.of(
                    QueryCondition.of(
                        Column.of(StoreConstants.ID),
                        group_id
                    )
                )
            )
        )

        if len(group_response.records) == 0:
            raise HTTPException(
                status_code=404,
                detail="No groups found with a matching id."
            )

        team_records: list[dict[str, Any]] = (
            self.__query_service.retrieve_records(
                QueryRequest(
                    table=Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        Team.TARGET_TABLE
                    ),
                    conditionGroup=QueryConditionGroup.of(
                        QueryCondition(
                            column=Column.of(StoreConstants.ID),
                            operator=ConditionOperator.IN,
                            value=team_ids
                        )
                    )
                )
            ).records
        )

        existing_ids: list[UUID] = list(
            map(
                lambda team:
                UUID(team[StoreConstants.ID]),
                team_records
            )
        )
        invalid_ids: list[UUID] = list(set(team_ids) - set(existing_ids))

        if len(invalid_ids) > 0:
            raise HTTPException(
                status_code=400,
                detail="No teams found with ids: " + str(
                    list(
                        map(
                            lambda team_id:
                            str(team_id),
                            invalid_ids
                        )
                    )
                )
            )

        group_team_response: QueryResponse = (
            self.__query_service.retrieve_records(
                QueryRequest(
                    table=Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        PredictorConstants.get_group_teams_table(tournament_id)
                    ),
                    conditionGroup=QueryConditionGroup.of(
                        QueryCondition.of(
                            Column.of("groupId"),
                            group_id
                        )
                    )
                )
            )
        )

        template_response: QueryResponse = (
            self.__query_service.retrieve_records(
                QueryRequest(
                    columns=[
                        Column.of("league", "teamsPerGroup")
                    ],
                    table=Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        Tournament.TARGET_TABLE,
                        "tourn"
                    ),
                    tableJoins=[
                        TableJoin.of(
                            Table.of(
                                PredictorConstants.PREDICTOR_SCHEMA,
                                TournamentTemplate.TARGET_TABLE,
                                "temp"
                            ),
                            QueryCondition.of(
                                Column.of("tourn", "templateId"),
                                Column.of("temp", StoreConstants.ID)
                            ),
                            TableJoinType.INNER
                        ),
                        TableJoin.of(
                            Table.of(
                                PredictorConstants.PREDICTOR_SCHEMA,
                                LeagueTemplate.TARGET_TABLE,
                                "league"
                            ),
                            QueryCondition.of(
                                Column.of("temp", "leagueTemplateId"),
                                Column.of("league", StoreConstants.ID)
                            ),
                            TableJoinType.INNER
                        )
                    ],
                    conditionGroup=QueryConditionGroup.of(
                        QueryCondition.of(
                            Column.of("tourn", StoreConstants.ID),
                            tournament_id
                        )
                    )
                )
            )
        )

        if (len(group_team_response.records) + len(team_ids)
                > template_response.records[0]["teamsPerGroup"]):
            raise HTTPException(
                status_code=409,
                detail="The number of groups in this tournament's group stage "
                       "has been exceeded."
            )

        self.__query_service.update_records(
            UpdateRequest(
                operation=SqlOperator.INSERT,
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    PredictorConstants.get_group_teams_table(tournament_id)
                ),
                records=list(
                    map(
                        lambda team_id:
                        {
                            "groupId": group_id,
                            "teamId": team_id
                        },
                        team_ids
                    )
                )
            )
        )

        return self.__retrieve_and_build_groups(
            tournament_id,
            QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of("group", StoreConstants.ID),
                    group_id
                )
            )
        )[0]

    def remove_team_from_group(
            self,
            tournament_id: UUID,
            group_id: UUID,
            team_id: UUID) -> Group:
        """
        Remove a single team from a group.

        Args:
            tournament_id (UUID): The id of the tournament the group belongs
                to.
            group_id (UUID): The id of the group to remove the team from.
            team_id (UUID): The id of the team to remove from the group.

        Return:
            Group: The updated group.
        """
        self.__tournament_service.get_tournament_by_id(tournament_id)

        if not self.__tournament_has_group_stage(tournament_id):
            raise HTTPException(
                status_code=404,
                detail="The tournament with the supplied id does not have a "
                       "group stage."
            )

        group_response: QueryResponse = self.__query_service.retrieve_records(
            QueryRequest(
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    Group.get_target_table(tournament_id)
                ),
                conditionGroup=QueryConditionGroup.of(
                    QueryCondition.of(
                        Column.of(StoreConstants.ID),
                        group_id
                    )
                )
            )
        )

        if len(group_response.records) == 0:
            raise HTTPException(
                status_code=404,
                detail="No groups found with a matching id."
            )

        self.__query_service.update_records(
            UpdateRequest(
                operation=SqlOperator.DELETE,
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    PredictorConstants.get_group_teams_table(tournament_id)
                ),
                conditionGroup=QueryConditionGroup.of(
                    QueryCondition.of(Column.of("groupId"), group_id),
                    QueryCondition.of(Column.of("teamId"), team_id)
                )
            )
        )

        return self.__retrieve_and_build_groups(
            tournament_id,
            QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of("group", StoreConstants.ID),
                    group_id
                )
            )
        )[0]

    def __tournament_has_group_stage(self, tournament_id: UUID) -> bool:
        """
        Determine whether a tournament has a group stage or not.

        Args:
            tournament_id (UUID): The id of the tournament.

        Returns:
            bool: True if the tournament has a group stage.
        """
        response: QueryResponse = self.__query_service.retrieve_records(
            QueryRequest(
                columns=[
                    Column.of("temp", "leagueTemplateId")
                ],
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    Tournament.TARGET_TABLE,
                    "tourn"
                ),
                tableJoins=[
                    TableJoin.of(
                        Table.of(
                            PredictorConstants.PREDICTOR_SCHEMA,
                            TournamentTemplate.TARGET_TABLE,
                            "temp"
                        ),
                        QueryCondition.of(
                            Column.of("tourn", "templateId"),
                            Column.of("temp", StoreConstants.ID)
                        ),
                        TableJoinType.INNER
                    )
                ],
                conditionGroup=QueryConditionGroup.of(
                    QueryCondition.of(
                        Column.of("tourn", StoreConstants.ID),
                        tournament_id
                    )
                )
            )
        )

        return response.records[0]["leagueTemplateId"] is not None

    def __retrieve_and_build_groups(
            self,
            tournament_id: UUID,
            condition_group: QueryConditionGroup = None) -> list[Group]:
        """
        Retrieve a set of groups and build them into objects containing their
        teams.

        Args:
             tournament_id (UUID): The id of the tournament the groups belong
                to.
             condition_group (QueryConditionGroup): The conditions by which
                to filter the groups by.

        Returns:
            list[Group]: The retrieved groups.
        """
        response: QueryResponse = self.__query_service.retrieve_records(
            QueryRequest(
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    Group.get_target_table(tournament_id),
                    "group"
                ),
                conditionGroup=condition_group,
                orderBy=[
                    OrderBy.of(Column.of("group", "name"))
                ]
            )
        )

        groups: list[Group] = list(
            map(
                lambda record:
                Group.model_validate(record),
                response.records
            )
        )

        team_response: QueryResponse = self.__query_service.retrieve_records(
            QueryRequest(
                columns=[
                    Column.of("gt", "groupId"),
                    Column.of("team", StoreConstants.ID),
                    Column.of("team", "name"),
                    Column.of("team", "imagePath"),
                    Column.of("team", "confederation")
                ],
                table=Table.of(
                    PredictorConstants.PREDICTOR_SCHEMA,
                    PredictorConstants.get_group_teams_table(tournament_id),
                    "gt"
                ),
                tableJoins=[
                    TableJoin.of(
                        Table.of(
                            PredictorConstants.PREDICTOR_SCHEMA,
                            Team.TARGET_TABLE,
                            "team"
                        ),
                        QueryCondition.of(
                            Column.of("gt", "teamId"),
                            Column.of("team", StoreConstants.ID)
                        ),
                        TableJoinType.INNER
                    ),
                    TableJoin.of(
                        Table.of(
                            PredictorConstants.PREDICTOR_SCHEMA,
                            Group.get_target_table(tournament_id),
                            "group"
                        ),
                        QueryCondition.of(
                            Column.of("gt", "groupId"),
                            Column.of("group", StoreConstants.ID)
                        ),
                        TableJoinType.INNER
                    )
                ],
                conditionGroup=condition_group,
                orderBy=[
                    OrderBy.of(
                        Column.of("team", "name")
                    )
                ]
            )
        )

        return list(
            map(
                lambda group:
                group.model_copy(update={
                    "teams": list(
                        map(
                            lambda filtered:
                            Team.model_validate(filtered),
                            list(
                                filter(
                                    lambda record:
                                    UUID(record["groupId"]) == group.id,
                                    team_response.records
                                )
                            )
                        )
                    )
                }),
                groups
            )
        )

File: src/main/java/com/kanemullett/service/GroupService.java
package com.kanemullett.service;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import com.kanemullett.model.Column;
import com.kanemullett.model.DatabaseRecord;
import com.kanemullett.model.Group;
import com.kanemullett.model.GroupRecord;
import com.kanemullett.model.GroupTeam;
import com.kanemullett.model.ImmutableGroup;
import com.kanemullett.model.ImmutableGroupTeam;
import com.kanemullett.model.ImmutableQueryCondition;
import com.kanemullett.model.ImmutableQueryRequest;
import com.kanemullett.model.ImmutableTableJoin;
import com.kanemullett.model.ImmutableTeam;
import com.kanemullett.model.ImmutableUpdateRequest;
import com.kanemullett.model.LeagueTemplate;
import com.kanemullett.model.QueryCondition;
import com.kanemullett.model.QueryConditionGroup;
import com.kanemullett.model.QueryRequest;
import com.kanemullett.model.QueryResponse;
import com.kanemullett.model.Table;
import com.kanemullett.model.TableJoin;
import com.kanemullett.model.Team;
import com.kanemullett.model.Tournament;
import com.kanemullett.model.TournamentTemplateRecord;
import com.kanemullett.model.UpdateRequest;
import com.kanemullett.model.type.ConditionOperator;
import com.kanemullett.model.type.Confederation;
import com.kanemullett.model.type.JoinType;
import com.kanemullett.model.type.SqlOperator;
import com.kanemullett.util.DatabaseConstants;
import com.kanemullett.util.PredictorConstants;

import jakarta.annotation.Nullable;

@Service
public class GroupService {

    private final DatabaseQueryService queryService;

    private static final String GROUPS_TABLE_ALIAS = "gr";
    private static final String GROUP_TEAMS_TABLE_ALIAS = "gt";
    private static final String LEAGUE_TEMPLATES_TABLE_ALIAS = "lt";
    private static final String TEAMS_TABLE_ALIAS = "te";
    private static final String TOURNAMENTS_TABLE_ALIAS = "to";
    private static final String TOURNAMENT_TEMPLATES_TABLE_ALIAS = "tt";

    public GroupService(DatabaseQueryService queryService) {
        this.queryService = queryService;
    }

    public List<Group> getGroups(String tournamentId) {

        verifyTournament(tournamentId);
        return retrieveAndBuildGroups(tournamentId, null);
    }

    public List<Group> updateGroups(String tournamentId, List<GroupRecord> groupUpdates) {
        verifyTournament(tournamentId);

        final UpdateRequest<GroupRecord> updateRequest = ImmutableUpdateRequest.<GroupRecord>builder()
            .operation(SqlOperator.UPDATE)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupRecord.getTargetTable(tournamentId)))
            .records(groupUpdates)
            .build();

        queryService.updateRecords(updateRequest);

        final QueryCondition queryCondition = ImmutableQueryCondition.builder()
            .column(Column.of(GROUP_TEAMS_TABLE_ALIAS, GroupTeam.GROUP_ID_COLUMN))
            .operator(ConditionOperator.IN)
            .value(groupUpdates.stream()
                .map(GroupRecord::getId)
                .toList())
            .build();

        return retrieveAndBuildGroups(tournamentId, QueryConditionGroup.of(queryCondition));
    }

    public Group getGroupById(String tournamentId, String groupId) {
        verifyTournament(tournamentId);
        verifyGroup(tournamentId, groupId);

        return retrieveAndBuildGroup(tournamentId, groupId);
    }

    public Group addTeamsToGroup(String tournamentId, String groupId, List<String> teamIds) {
        verifyTournament(tournamentId);
        verifyGroup(tournamentId, groupId);

        final QueryRequest<Team> teamRequest = ImmutableQueryRequest.<Team>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, Team.TARGET_TABLE))
            .conditionGroup(QueryConditionGroup.of(
                ImmutableQueryCondition.builder()
                    .column(Column.of(DatabaseConstants.ID))
                    .operator(ConditionOperator.IN)
                    .value(teamIds)
                    .build()
            ))
            .recordClass(Team.class)
            .build();

        final QueryResponse<Team> teamResponse = queryService.retrieveRecords(teamRequest);

        final List<String> existingIds = teamResponse.getRecords().stream()
            .map(Team::getId)
            .toList();

        final List<String> invalidIds = teamIds.stream()
            .filter(teamId -> !existingIds.contains(teamId))
            .toList();

        if (!invalidIds.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "No teams found with ids: " + String.join(", ", invalidIds));
        }

        final QueryRequest<GroupTeam> groupTeamsRequest = ImmutableQueryRequest.<GroupTeam>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupTeam.getTargetTable(tournamentId)))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(GroupTeam.GROUP_ID_COLUMN),
                    groupId
                )
            ))
            .recordClass(GroupTeam.class)
            .build();

        final QueryResponse<GroupTeam> groupTeamsResponse = queryService.retrieveRecords(groupTeamsRequest);
        final int teamsInGroup = groupTeamsResponse.getRecordCount();

        final QueryRequest<DatabaseRecord> teamsPerGroupRequest = ImmutableQueryRequest.<DatabaseRecord>builder()
            .columns(List.of(
                Column.of(LEAGUE_TEMPLATES_TABLE_ALIAS, LeagueTemplate.TEAMS_PER_GROUP_COLUMN)
            ))
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, Tournament.TARGET_TABLE, TOURNAMENTS_TABLE_ALIAS))
            .joins(List.of(
                TableJoin.of(
                    Table.of(PredictorConstants.PREDICTOR_SCHEMA, TournamentTemplateRecord.TARGET_TABLE, TOURNAMENT_TEMPLATES_TABLE_ALIAS),
                    QueryCondition.of(
                        Column.of(TOURNAMENTS_TABLE_ALIAS, Tournament.TEMPLATE_ID_COLUMN),
                        Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, DatabaseConstants.ID)
                    ),
                    JoinType.INNER
                ),
                TableJoin.of(
                    Table.of(PredictorConstants.PREDICTOR_SCHEMA, LeagueTemplate.TARGET_TABLE, LEAGUE_TEMPLATES_TABLE_ALIAS),
                    QueryCondition.of(
                        Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, TournamentTemplateRecord.LEAGUE_TEMPLATE_ID_COLUMN),
                        Column.of(LEAGUE_TEMPLATES_TABLE_ALIAS, DatabaseConstants.ID)
                    ),
                    JoinType.INNER
                )
            ))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(TOURNAMENTS_TABLE_ALIAS, DatabaseConstants.ID),
                    tournamentId
                )
            ))
            .recordClass(DatabaseRecord.class)
            .build();

        final QueryResponse<DatabaseRecord> teamsPerGroupResponse = queryService.retrieveRecords(teamsPerGroupRequest);
        final int teamsPerGroup = teamsPerGroupResponse.getRecordCount();

        if (teamsInGroup + teamIds.size() > teamsPerGroup) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Adding teams to the specified group will exceed its maximum team count.");
        }

        final UpdateRequest<GroupTeam> updateRequest = ImmutableUpdateRequest.<GroupTeam>builder()
            .operation(SqlOperator.INSERT)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupTeam.getTargetTable(tournamentId)))
            .records(teamIds.stream()
                .map(teamId -> ImmutableGroupTeam.builder()
                    .groupId(groupId)
                    .teamId(teamId)
                    .build())
                .toList())
            .build();

        queryService.updateRecords(updateRequest);

        return retrieveAndBuildGroup(tournamentId, groupId);
    }

    public Group removeTeamFromGroup(String tournamentId, String groupId, String teamId) {
        verifyTournament(tournamentId);
        verifyGroup(tournamentId, groupId);

        final UpdateRequest<GroupTeam> updateRequest = ImmutableUpdateRequest.<GroupTeam>builder()
            .operation(SqlOperator.DELETE)
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupTeam.getTargetTable(tournamentId)))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(GroupTeam.GROUP_ID_COLUMN),
                    groupId
                ),
                QueryCondition.of(
                    Column.of(GroupTeam.TEAM_ID_COLUMN),
                    teamId
                )
            ))
            .build();

        queryService.updateRecords(updateRequest);

        return retrieveAndBuildGroup(tournamentId, groupId);
    }

    private void verifyTournament(String tournamentId) {
        final QueryRequest<DatabaseRecord> queryRequest = ImmutableQueryRequest.<DatabaseRecord>builder()
            .columns(List.of(
                Column.of(TOURNAMENTS_TABLE_ALIAS, DatabaseConstants.ID),
                Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, TournamentTemplateRecord.LEAGUE_TEMPLATE_ID_COLUMN)
            ))
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, Tournament.TARGET_TABLE, TOURNAMENTS_TABLE_ALIAS))
            .joins(List.of(ImmutableTableJoin.builder()
                    .table(Table.of(
                        PredictorConstants.PREDICTOR_SCHEMA,
                        TournamentTemplateRecord.TARGET_TABLE,
                        TOURNAMENT_TEMPLATES_TABLE_ALIAS
                    ))
                    .joinType(JoinType.INNER)
                    .joinCondition(QueryCondition.of(
                        Column.of(TOURNAMENTS_TABLE_ALIAS, Tournament.TEMPLATE_ID_COLUMN),
                        Column.of(TOURNAMENT_TEMPLATES_TABLE_ALIAS, DatabaseConstants.ID)
                    ))
                    .build()
            ))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(TOURNAMENTS_TABLE_ALIAS, DatabaseConstants.ID),
                    tournamentId
                )
            ))
            .recordClass(DatabaseRecord.class)
            .build();

        final QueryResponse<DatabaseRecord> queryResponse = queryService.retrieveRecords(queryRequest);

        if (queryResponse.getRecordCount() == 0) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "No tournaments found with a matching id.");
        }

        final DatabaseRecord tournamentRecord = queryResponse.getRecords().get(0);

        if (tournamentRecord.getData().get(TournamentTemplateRecord.LEAGUE_TEMPLATE_ID_COLUMN) == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "The tournament with the supplied id does not have a group stage.");
        }
    }

    private void verifyGroup(String tournamentId, String groupId) {
        final QueryRequest<GroupRecord> queryRequest = ImmutableQueryRequest.<GroupRecord>builder()
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupRecord.getTargetTable(tournamentId)))
            .conditionGroup(QueryConditionGroup.of(
                QueryCondition.of(
                    Column.of(DatabaseConstants.ID),
                    groupId
                )
            ))
            .recordClass(GroupRecord.class)
            .build();

        final QueryResponse<GroupRecord> queryResponse = queryService.retrieveRecords(queryRequest);

        if (queryResponse.getRecordCount() == 0) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "No groups found with a matching id.");
        }
    }

    private Group retrieveAndBuildGroup(String tournamentId, String groupId) {
        final QueryConditionGroup conditionGroup = QueryConditionGroup.of(
            QueryCondition.of(
                Column.of(GROUP_TEAMS_TABLE_ALIAS, GroupTeam.GROUP_ID_COLUMN),
                groupId
            )
        );

        return retrieveAndBuildGroups(tournamentId, conditionGroup).get(0);
    }

    private List<Group> retrieveAndBuildGroups(String tournamentId, @Nullable QueryConditionGroup conditionGroup) {

        final QueryRequest<DatabaseRecord> queryRequest = ImmutableQueryRequest.<DatabaseRecord>builder()
            .columns(List.of(
                Column.of(GROUP_TEAMS_TABLE_ALIAS, GroupTeam.GROUP_ID_COLUMN),
                Column.of(GROUPS_TABLE_ALIAS, GroupRecord.NAME_COLUMN, "groupName"),
                Column.of(GROUP_TEAMS_TABLE_ALIAS, GroupTeam.TEAM_ID_COLUMN),
                Column.of(TEAMS_TABLE_ALIAS, Team.NAME_COLUMN, "teamName"),
                Column.of(TEAMS_TABLE_ALIAS, Team.IMAGE_PATH_COLUMN),
                Column.of(TEAMS_TABLE_ALIAS, Team.CONFEDERATION_COLUMN),
                Column.of(TEAMS_TABLE_ALIAS, Team.RANKING_COLUMN)
            ))
            .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupTeam.getTargetTable(tournamentId), GROUP_TEAMS_TABLE_ALIAS))
            .joins(List.of(
                ImmutableTableJoin.builder()
                    .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, GroupRecord.getTargetTable(tournamentId), GROUPS_TABLE_ALIAS))
                    .joinType(JoinType.INNER)
                    .joinCondition(QueryCondition.of(
                        Column.of(GROUP_TEAMS_TABLE_ALIAS, GroupTeam.GROUP_ID_COLUMN),
                        Column.of(GROUPS_TABLE_ALIAS, DatabaseConstants.ID)
                    ))
                    .build(),
                ImmutableTableJoin.builder()
                    .table(Table.of(PredictorConstants.PREDICTOR_SCHEMA, Team.TARGET_TABLE, TEAMS_TABLE_ALIAS))
                    .joinType(JoinType.INNER)
                    .joinCondition(QueryCondition.of(
                        Column.of(GROUP_TEAMS_TABLE_ALIAS, GroupTeam.TEAM_ID_COLUMN),
                        Column.of(TEAMS_TABLE_ALIAS, DatabaseConstants.ID)
                    ))
                    .build()
            ))
            .conditionGroup(conditionGroup)
            .recordClass(DatabaseRecord.class)
            .build();

        final QueryResponse<DatabaseRecord> queryResponse = queryService.retrieveRecords(queryRequest);

        return queryResponse.getRecords().stream()
            .collect(Collectors.groupingBy(record -> (String) record.getData().get(GroupTeam.GROUP_ID_COLUMN), LinkedHashMap::new, Collectors.toList()))
            .entrySet().stream()
            .map(entry -> buildGroup(entry.getKey(), entry.getValue()))
            .collect(Collectors.toList());
    }

    private Group buildGroup(String groupId, List<DatabaseRecord> groupRecords) {
        final String groupName = (String) groupRecords.get(0).getData().get("groupName");

        final List<Team> teams = groupRecords.stream()
            .map(record -> ImmutableTeam.builder()
                .id((String) record.getData().get(GroupTeam.TEAM_ID_COLUMN))
                .name((String) record.getData().get("teamName"))
                .imagePath((String) record.getData().get(Team.IMAGE_PATH_COLUMN))
                .confederation(Enum.valueOf(Confederation.class, (String) record.getData().get(Team.CONFEDERATION_COLUMN)))
                .ranking((Integer) record.getData().get(Team.RANKING_COLUMN))
                .build())
            .collect(Collectors.toList());

        return ImmutableGroup.builder()
            .id(groupId)
            .name(groupName)
            .teams(teams)
            .build();
    }
}
